### PR TITLE
lib: lte_link_control: add %UICCPOWERSAVE

### PR DESCRIPTION
**Pull request by MirkoCovizzi**
Adds support for the `UICCPOWERSAVE` AT command.

### CODE REVIEW

File: samples/cellular/modem_shell/src/link/link_shell.c [hunk 1]
```diff
@@ -1997,6 +2043,85 @@ static int link_shell_redmob(const struct shell *shell, size_t argc, char **argv
 	return 0;
 }
 
+static int link_shell_uiccpowersave(const struct shell *shell, size_t argc, char **argv)
```
Comment (trantanen) [inline samples/cellular/modem_shell/src/link/link_shell.c:2046]: Keep alphabetical order of subcommand functions

File: samples/cellular/modem_shell/src/link/link_shell.c [hunk 2]
```diff
@@ -2664,6 +2789,10 @@ SHELL_STATIC_SUBCMD_SET_CREATE(
 		tau, NULL,
 		"Subscribe/unsubscribe for modem TAU pre-warning notifications.",
 		link_shell_tau, 0, 10),
+	SHELL_CMD_ARG(
```
Comment (trantanen) [inline samples/cellular/modem_shell/src/link/link_shell.c:2792]: Keep alphabetical order.

File: samples/cellular/modem_shell/src/link/link_shell.c [hunk 3]
```diff
@@ -1997,6 +2043,85 @@ static int link_shell_redmob(const struct shell *shell, size_t argc, char **argv
...
+	char snum[10];
+
+	optreset = 1;
+	optind = 1;
+	int opt;
+
+	while ((opt = getopt_long(argc, argv, short_options, long_options, NULL)) != -1) {
+		switch (opt) {
+		case 'r':
+			if (operation != LINK_OPERATION_NONE) {
```
Comment (trantanen) [inline samples/cellular/modem_shell/src/link/link_shell.c:2060]: This branch should be unnecessary. Just one of the operations returned by link_shell_getopt_operation() is possible so you cannot really get here.

File: samples/cellular/modem_shell/src/link/link_shell.c [hunk 4]
```diff
@@ -1997,6 +2043,85 @@ static int link_shell_redmob(const struct shell *shell, size_t argc, char **argv
...
+	int ret;
+	enum link_shell_operation operation = LINK_OPERATION_NONE;
+	enum lte_lc_uiccpowersave_mode uiccpowersave_mode = LINK_UICCPOWERSAVE_NONE;
+	char snum[10];
+
+	optreset = 1;
+	optind = 1;
+	int opt;
+
+	while ((opt = getopt_long(argc, argv, short_options, long_options, NULL)) != -1) {
```
Comment (trantanen) [inline samples/cellular/modem_shell/src/link/link_shell.c:2057]: Since there is just one argument, should we remove while loop?

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 5]
```diff
@@ -686,6 +687,19 @@ Modem libraries
...
+    * The :c:enum:`LTE_LC_UICCPOWERSAVE_BYPASS` mode. By setting this mode, UICC is deactivated by bypassing certain conditions. This mode is supported by Modem Firmware 2.0.2 or higher.
+    * The :c:enum:`LTE_LC_UICCPOWERSAVE_RRC` mode. By setting this mode, UICC is deactivated by bypassing certain conditions and when entering RRC IDLE state. This mode is supported by Modem Firmware 2.0.2 or higher.
+    * The :kconfig:option:`CONFIG_LTE_UICCPOWERSAVE_DEFAULT` Kconfig option for setting the :c:enum:`LTE_LC_UICCPOWERSAVE_DEFAULT` mode during initialization.
+    * The :kconfig:option:`CONFIG_LTE_UICCPOWERSAVE_POLLING` Kconfig option for setting the :c:enum:`LTE_LC_UICCPOWERSAVE_POLLING` mode during initialization.
+    * The :kconfig:option:`CONFIG_LTE_UICCPOWERSAVE_BYPASS` Kconfig option for setting the :c:enum:`LTE_LC_UICCPOWERSAVE_BYPASS` mode during initialization.
+    * The :kconfig:option:`CONFIG_LTE_UICCPOWERSAVE_RRC` Kconfig option for setting the :c:enum:`LTE_LC_UICCPOWERSAVE_RRC` mode during initialization.
+
```
[the remark this answers is not in the document]
Comment (trantanen) [reply doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst]: Agreed.

File: include/modem/lte_lc.h [hunk 6]
```diff
@@ -651,6 +651,47 @@ enum lte_lc_reduced_mobility_mode {
...
+	 */
+	LTE_LC_UICCPOWERSAVE_POLLING = 1,
+
+	/**
+	 * Bypass conditions related to EF-AD, PIN state, and minimum deactivation time when
+	 * deactivating the UICC.
+	 * After deactivation on eDRX sleep, the UICC is activated only when the EMM (EPS
+	 * Mobility Management) is starting to initiate signalling to the network.
+	 */
+	LTE_LC_UICCPOWERSAVE_BYPASS = 2,
```
[the remark this answers is not in the document]
Comment (MirkoCovizzi) [reply include/modem/lte_lc.h:686]: I tried that in the first implementation, but it wouldn't be as intuitive to remember for the user.

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (MirkoCovizzi) [reply doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst]: This was updated before your review. Please refer to that version.

File: include/modem/lte_lc.h [hunk 7]
```diff
@@ -651,6 +651,47 @@ enum lte_lc_reduced_mobility_mode {
...
+enum lte_lc_uiccpowersave_mode {
+	/**
+	 * Default functionality.
+	 * The UICC is deactivated during eDRX only when:
+	 * - \li It is allowed by the EF-AD (Elementary File - Administrative Data)
+	 * - \li Application PIN is disabled
+	 * - \li Sleep mode time is longer than the minimum time limit (1 minute in ultra
+	 *   low power mode and 5 minutes in other power modes)
+	 *
+	 * @note AT&T has a shorter minimum time of 20.48 seconds.
```
[the remark this answers is not in the document]
Comment (MirkoCovizzi) [reply include/modem/lte_lc.h:666]: @trantanen

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 8]
```diff
@@ -686,6 +687,12 @@ Modem libraries
     * To use the :ref:`at_parser_readme` library instead of the :ref:`at_cmd_parser_readme` library.
     * The :c:func:`lte_lc_neighbor_cell_measurement` function to return an error for invalid GCI count.
 
+  * Added:
+
+    * The :c:func:`lte_lc_uiccpowersave_get` function for reading the current UICC power saving mode, and the :c:func:`lte_lc_uiccpowersave_set` function for setting the UICC power saving mode. These functions are supported by Modem Firmware 2.0.2 or higher. Please refer to the :ref:`lte_lc_readme` documentation for more information.
+    * The :c:enum:`LTE_LC_UICCPOWERSAVE_DEFAULT`, :c:enum:`LTE_LC_UICCPOWERSAVE_POLLING`, :c:enum:`LTE_LC_UICCPOWERSAVE_BYPASS`, and :c:enum:`LTE_LC_UICCPOWERSAVE_RRC` UICC power saving modes. These modes are supported by Modem Firmware 2.0.2 or higher. Please refer to the :ref:`lte_lc_readme` documentation for more information.
```
Comment (eivindj-nordic) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:693]: I think this entry is not needed as it is referenced by/related to the above entry.

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 9]
```diff
@@ -686,6 +687,12 @@ Modem libraries
...
+    * The :kconfig:option:`CONFIG_LTE_UICCPOWERSAVE_DEFAULT`, :kconfig:option:`CONFIG_LTE_UICCPOWERSAVE_POLLING`, :kconfig:option:`CONFIG_LTE_UICCPOWERSAVE_BYPASS`, and :kconfig:option:`CONFIG_LTE_UICCPOWERSAVE_RRC` Kconfig options for setting the respective UICC power saving modes during initialization. These Kconfig options are supported by Modem Firmware 2.0.2 or higher.
```
Comment (eivindj-nordic) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:694]: I think we can reference the choice instead of each option
```suggestion
   * The :kconfig:option:`LTE_UICCPOWERSAVE_MODE` Kconfig choice to set the UICC power saving mode during initialization. This is supported by Modem Firmware 2.0.2 and higher.
```

File: include/modem/lte_lc.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (tokangas) [reply include/modem/lte_lc.h:666]: I would leave it out from the SDK, these kind of exceptions are probably better to be described in the modem documentation.

File: include/modem/lte_lc.h [hunk 10]
```diff
@@ -651,6 +651,47 @@ enum lte_lc_reduced_mobility_mode {
...
+	 *
+	 * Proactive polling interval can be modified during eDRX sleep only if allowed in EF-AD.
+	 */
+	LTE_LC_UICCPOWERSAVE_DEFAULT = 0,
```
[the remark this answers is not in the document]
Comment (tokangas) [reply include/modem/lte_lc.h:670]: I have mixed feelings about this. 😄 I don't have a strong opinion.

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 11]
```diff
@@ -368,6 +368,7 @@ Cellular samples
 * :ref:`modem_shell_application` sample:
 
   * Updated to use the :ref:`at_parser_readme` library instead of the :ref:`at_cmd_parser_readme` library.
+  * Added the ``uiccpowersave`` link command for reading/setting UICC power saving mode. This feature is supported by Modem Firmware 2.0.2 or higher.
```
Comment (divipillai) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:371]:
```suggestion
  * Added the ``uiccpowersave`` link command for reading or setting UICC power saving mode. 
    This feature is supported by modem firmware v2.0.2 or higher.
  * Updated to use the :ref:`at_parser_readme` library instead of the :ref:`at_cmd_parser_readme` library.
```
line no: 370 and 371 could be interchanged to follow the order Added, updated.

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 12]
```diff
@@ -686,6 +687,12 @@ Modem libraries
     * To use the :ref:`at_parser_readme` library instead of the :ref:`at_cmd_parser_readme` library.
     * The :c:func:`lte_lc_neighbor_cell_measurement` function to return an error for invalid GCI count.
 
+  * Added:
+
+    * The :c:func:`lte_lc_uiccpowersave_get` function for reading the current UICC power saving mode, and the :c:func:`lte_lc_uiccpowersave_set` function for setting the UICC power saving mode. These functions are supported by Modem Firmware 2.0.2 or higher. Please refer to the :ref:`lte_lc_readme` documentation for more information.
```
Comment (divipillai) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:692]:
```suggestion
    * The :c:func:`lte_lc_uiccpowersave_get` function for reading the current UICC power saving mode, and the :c:func:`lte_lc_uiccpowersave_set` function for setting the UICC power saving mode. 
      These functions are supported by modem firmware v2.0.2 or higher. 
      Refer to the :ref:`lte_lc_readme` documentation for more information.
```

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst (hunk 8 above)
Comment (divipillai) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:693]:
```suggestion
    * The :c:enum:`LTE_LC_UICCPOWERSAVE_DEFAULT`, :c:enum:`LTE_LC_UICCPOWERSAVE_POLLING`, :c:enum:`LTE_LC_UICCPOWERSAVE_BYPASS`, and :c:enum:`LTE_LC_UICCPOWERSAVE_RRC` UICC power saving modes.
      These modes are supported by modem Firmware v2.0.2 or higher. 
      Refer to the :ref:`lte_lc_readme` documentation for more information.
```

File: include/modem/lte_lc.h [hunk 13]
```diff
@@ -651,6 +651,47 @@ enum lte_lc_reduced_mobility_mode {
 	LTE_LC_REDUCED_MOBILITY_DISABLED = 2,
 };
 
+/**
+ * UICC power saving mode.
```
Comment (tokangas) [inline include/modem/lte_lc.h:655]: We should add this text somewhere. Maybe it's best to add it here, so don't need to have it in both functions.
```suggestion
 * UICC power saving mode.
 *
 * Affects only UICC deactivation and suspension during eDRX and idle state. PSM functionality
 * is not affected.
```

File: include/modem/lte_lc.h [hunk 14]
```diff
@@ -1740,6 +1781,35 @@ int lte_lc_reduced_mobility_get(enum lte_lc_reduced_mobility_mode *mode);
...
+ * Set the UICC power saving mode.
+ *
+ * @note This feature is only supported by modem firmware versions >= 2.0.2.
+ *
+ * @param[in] mode The UICC power saving mode to set.
+ *
+ * @retval 0 if the new UICC power saving mode was successfully set.
+ * @retval -EFAULT if an AT command failed to execute.
+ *
+ * @note This function configures the modem to use a specific UICC power saving mode.
```
Comment (tokangas) [inline include/modem/lte_lc.h:1809]: Same here, I don't think this note is needed.

File: lib/lte_link_control/Kconfig [hunk 15]
```diff
@@ -349,6 +349,57 @@ config LTE_LC_WORKQUEUE_STACK_SIZE
 	default 1280 if LOG_MODE_IMMEDIATE
 	default 1024
 
+choice LTE_UICCPOWERSAVE_MODE
```
[the remark this answers is not in the document]
Comment (tokangas) [reply lib/lte_link_control/Kconfig:352]: I'll confirm from the modem team whether the configuration is stored into NVM or not. At least it's not mentioned in the AT command specification. If it would be stored, one option would be that "default" does nothing, i.e. uses the configuration from modem NVM.

File: lib/lte_link_control/lte_lc_modem_hooks.c [hunk 16]
```diff
@@ -87,6 +87,12 @@ static void on_modem_init(int err, void *ctx)
 	(void)nrf_modem_at_printf("AT%%FEACONF=0,3,%d",
 		IS_ENABLED(CONFIG_LTE_PLMN_SELECTION_OPTIMIZATION));
 
+	/* Only supported in mfw 2.0.2 and newer.
+	 * Ignore the return value; an error likely means that the feature
+	 * is not supported.
+	 */
+	(void)lte_lc_uiccpowersave_set(CONFIG_LTE_UICCPOWERSAVE_MODE_VALUE);
```
[the remark this answers is not in the document]
Comment (tokangas) [reply lib/lte_link_control/lte_lc_modem_hooks.c:94]: It would be easiest to just send the AT command directly. On the other hand, if the configuration is not saved into modem NVM, we could call `lte_lc_uiccpowersave_set()`, but only when the configured mode is something else than default. In that case the error would be appropriate, because it would only be logged when the user tries to enable UICC power saving mode with MFW which does not support it.

File: samples/cellular/modem_shell/src/link/link_shell.c [hunk 17]
```diff
@@ -357,6 +358,40 @@ static const char link_propripsm_usage_str[] =
 	"  -e, --enable,       Enable proprietary PSM\n"
 	"  -h, --help,         Shows this help information";
 
+static const char link_uiccpowersave_usage_str[] =
+	"Usage: link uiccpowersave [options] | --read\n"
+	"Options:\n"
+	"  -r, --read,         Read and print the current UICC power saving mode.\n"
+	"      --default_mode, Default functionality.\n"
```
Comment (tokangas) [inline samples/cellular/modem_shell/src/link/link_shell.c:365]: Let's align the option names when it has been decided how the different modes are named.

File: samples/cellular/modem_shell/src/link/link_shell.c [hunk 18]
```diff
@@ -357,6 +358,40 @@ static const char link_propripsm_usage_str[] =
...
+	"                      The UICC is deactivated during eDRX only when:\n"
+	"                      - It is allowed by the EF-AD (Elementary File - Administrative\n"
+	"                        Data)\n"
+	"                      - Application PIN is disabled\n"
+	"                      - Sleep mode time is longer than the minimum time limit (1 minute\n"
+	"                        in ultra low power mode and 5 minutes in other power modes)\n"
+	"                      Note: AT&T has a shorter minimum time of 20.48 seconds.\n"
+	"                      Proactive polling interval can be modified during eDRX sleep only\n"
+	"                      if allowed in EF-AD.\n"
+	"                      Supported with modem firmware versions >= 2.0.2.\n"
```
Comment (tokangas) [inline samples/cellular/modem_shell/src/link/link_shell.c:375]: Instead of repeating the version requirement for each option, we should rather have one note for the whole command. If MFW doesn't support the feature, also `--read` would fail.

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst (hunk 11 above)
[the remark this answers is not in the document]
Comment (divipillai) [reply doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:371]: It is okay to mention MFW version limitation. For example, it is mentioned here in the "nRF Cloud" section. https://docs.nordicsemi.com/bundle/ncs-latest/page/nrf/releases_and_maturity/releases/release-notes-2.7.0.html#highlights.

Agree to keep the "Modem Firmware" to lower case and adding 'v' prefix to the version number. This was one of the suggestion from my side along with other suggestions.

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst (hunk 9 above)
Comment (divipillai) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:694]:
```suggestion
    * The :kconfig:option:`CONFIG_LTE_UICCPOWERSAVE_MODE` Kconfig choice to set the UICC power saving mode during initialization. 
      This is supported by Modem Firmware 2.0.2 and higher.
```

File: include/modem/lte_lc.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (MirkoCovizzi) [reply include/modem/lte_lc.h:666]: @tokangas

[the remark this answers is not in the document]
Comment (trantanen) [reply include/modem/lte_lc.h:666]: I'd say no. https://github.com/nrfconnect/sdk-nrf/pull/17335/files#r1762675122

File: include/modem/lte_lc.h [hunk 19]
```diff
@@ -651,6 +651,47 @@ enum lte_lc_reduced_mobility_mode {
...
+	 * match the eDRX sleep time regardless of EF-AD content.
+	 * If the interval is already longer than the sleep time, it is not affected.
+	 */
+	LTE_LC_UICCPOWERSAVE_POLLING = 1,
+
+	/**
+	 * Bypass conditions related to EF-AD, PIN state, and minimum deactivation time when
+	 * deactivating the UICC.
+	 * After deactivation on eDRX sleep, the UICC is activated only when the EMM (EPS
+	 * Mobility Management) is starting to initiate signalling to the network.
```
Comment (divipillai) [inline include/modem/lte_lc.h:684]:
```suggestion
	 * Mobility Management) is starting to initiate signaling to the network.
```

File: include/modem/lte_lc.h [hunk 20]
```diff
@@ -651,6 +651,47 @@ enum lte_lc_reduced_mobility_mode {
...
+ * UICC power saving mode.
+ */
+enum lte_lc_uiccpowersave_mode {
+	/**
+	 * Default functionality.
+	 * The UICC is deactivated during eDRX only when:
+	 * - \li It is allowed by the EF-AD (Elementary File - Administrative Data)
+	 * - \li Application PIN is disabled
+	 * - \li Sleep mode time is longer than the minimum time limit (1 minute in ultra
+	 *   low power mode and 5 minutes in other power modes)
```
Comment (divipillai) [inline include/modem/lte_lc.h:664]:
```suggestion
	 * The UICC is deactivated during eDRX only when:
	 * \li It is allowed by the EF-AD (Elementary File - Administrative Data)
	 * \li Application PIN is disabled
	 * \li Sleep mode time is longer than the minimum time limit (1 minute in ultra
	 * low power mode and 5 minutes in other power modes)
```

File: lib/lte_link_control/Kconfig [hunk 21]
```diff
@@ -349,6 +349,57 @@ config LTE_LC_WORKQUEUE_STACK_SIZE
...
+	prompt "UICC power saving mode."
+	default LTE_UICCPOWERSAVE_DEFAULT
+	help
+	  Select the UICC power saving mode.
+	  Note: This feature is supported by Modem Firmware 2.0.2 or higher.
```
Comment (divipillai) [inline lib/lte_link_control/Kconfig:357]:
```suggestion
	  Note: This feature is supported by modem firmware 2.0.2 or higher.
```

**Review by divipillai — COMMENTED** [review]
Some minor suggestions.

File: include/modem/lte_lc.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (trantanen) [reply include/modem/lte_lc.h:686]: Neither solution being perfect for default, I think I'd vote for consistency and have `MODE_0`. Just because `MODE_DEFAULT` is not some well-known default functionality that people could understand without reading the description.
Also, normally the default value is just one of the values which doesn't have wording "default" in it although we do have those "default" ones too.

File: include/modem/lte_lc.h (hunk 13 above)
Comment (trantanen) [reply include/modem/lte_lc.h:655]: I'd reword your suggestion to avoid confusion that this affects UICC deactivation and suspension functionality. I assume it's more about those being used to achieve power savings in eDRX and idle states.

File: lib/lte_link_control/Kconfig (hunk 15 above)
[the remark this answers is not in the document]
Comment (tokangas) [reply lib/lte_link_control/Kconfig:352]: I confirmed from the modem team, that `%UICCPOWERSAVE` setting is not saved into NVM, so it should be always disabled/default/0 when the modem starts.

File: include/modem/lte_lc.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (trantanen) [reply include/modem/lte_lc.h:686]: Discussed this with @tokangas. Because `MODE_0` is default functionality of the modem, I'm open for `MODE_DEFAULT`. I had somehow assumed that there would be some other "no power saving state" if this AT command is not used.

File: lib/lte_link_control/lte_lc.c [hunk 22]
```diff
@@ -1821,6 +1821,39 @@ int lte_lc_reduced_mobility_set(enum lte_lc_reduced_mobility_mode mode)
...
+int lte_lc_uiccpowersave_get(enum lte_lc_uiccpowersave_mode *mode)
+{
+	int ret;
+	uint16_t mode_tmp;
+
+	if (mode == NULL) {
+		return -EINVAL;
+	}
+
+	ret = nrf_modem_at_scanf("AT%UICCPOWERSAVE?", "%%UICCPOWERSAVE: %hu", &mode_tmp);
```
Comment (trantanen) [inline lib/lte_link_control/lte_lc.c:1833]: I'm adding here a fundamental question. Do we need these functions for `AT%UICCPOWERSAVE` command at all? Or do you have some strategy which AT commands are wrapped into functions and which are not because we do not have functions for all AT commands.

I think the obvious AT commands are such that have complex syntax either in command, response, or notification. And those that have the notification in the first place. This `AT%UICCPOWERSAVE` doesn't belong into those categories but there may be other reasons to still do it.

Comment (tokangas) [reply lib/lte_link_control/lte_lc.c:1833]: Even without LTE LC support, it's easy for an application to enable `%UICCPOWERSAVE`. I think pretty much the only benefit of having it in LTE LC is that it makes it more visible, the user doesn't have to read through the AT command specification to find it. I'm not sure this is good enough reason to have it in LTE LC, though. Especially, when we have two different ways to apply this simple configuration, a Kconfig option and a function call (actually two, because there's also a read function).

File: include/modem/lte_lc.h (hunk 13 above)
Comment (MirkoCovizzi) [reply include/modem/lte_lc.h:655]: > I'd reword your suggestion to avoid confusion that this affects UICC deactivation and suspension functionality. I assume it's more about those being used to achieve power savings in eDRX and idle states.

What would your rewrite it to? I don't understand the nuances of how it's different from @tokangas version.

File: lib/lte_link_control/lte_lc.c (hunk 22 above)
Comment (trantanen) [reply lib/lte_link_control/lte_lc.c:1833]: > I think that it is one of those functionalities that has potentially immediate impact for most applications, hence it can be setup in the LTE link controller for all the applications/samples to benefit from.

Hmmm. mode 0 is the default behavior of the modem before this command was ever introduced. This command brings in additional modes (1-3) that does additional power savings in some use cases such as eDRX. So use of this AT command doesn't have much benefit for most applications. I'm not expecting this to be used much but there are certainly some customers who wants to look into it. And they might just as well use the AT command directly with their preferred mode.
